grammar.js

```diff
@@ -1247,6 +1247,60 @@ module.exports = grammar(C, {
       repeat1(choice($.raw_string_literal, $.string_literal)),
     ),
 
+    _integer_literal: $ => {
+      const separator = '\'';
+      const binary = /[01]/;
+      const binaryDigits = seq(repeat1(binary), repeat(seq(separator, repeat1(binary))));
+      const decimal = /[0-9]/;
+      const firstDecimal = /[1-9]/;
+      const decimalDigits = seq(firstDecimal, repeat(decimal), repeat(seq(separator, repeat1(decimal))));
+      const hex = /[0-9a-fA-F]/;
+      const hexDigits = seq(repeat1(hex), repeat(seq(separator, repeat1(hex))));
+      const octal = /[0-7]/;
+      const octalDigits = seq('0', repeat(octal), repeat(seq(separator, repeat1(octal))));
+      const suffix = /(ll|LL)[uU]?|[uU](ll|LL)?|[uU][lL]?|[uU][zZ]?|[lL][uU]?|[zZ][uU]?/;
+      return token(seq(
+        optional(/[-\+]/),
+        choice(
+          seq(choice('0b', '0B'), binaryDigits),
+          decimalDigits,
+          seq(choice('0x', '0X'), hexDigits),
+          octalDigits,
+        ),
+        optional(suffix),
+      ));
+    },
+
+    _float_literal: $ => {
+      const sign = /[-\+]/;
+      const separator = '\'';
+      const hex = /[0-9a-fA-F]/;
+      const decimal = /[0-9]/;
+      const hexDigits = seq(repeat1(hex), repeat(seq(separator, repeat1(hex))));
+      const decimalDigits = seq(repeat1(decimal), repeat(seq(separator, repeat1(decimal))));
+      const hexExponent = seq(/[pP]/, optional(sign), decimalDigits);
+      const decimalExponent = seq(/[eE]/, optional(sign), decimalDigits);
+      const suffix = /([fF](16|32|64|128)?)|[lL]|(bf16|BF16)/;
+      return token(seq(
+        optional(sign),
+        choice(
+          seq(decimalDigits, decimalExponent),
+          seq(decimalDigits, '.', optional(decimalDigits), optional(decimalExponent)),
+          seq('.', decimalDigits, optional(decimalExponent)),
+          seq(
+            choice('0x', '0X'),
+            choice(
+              hexDigits,
+              seq(hexDigits, '.', optional(hexDigits)),
+              seq('.', hexDigits)),
+            hexExponent),
+        ),
+        optional(suffix),
+      ));
+    },
+
+    number_literal: $ => choice($._integer_literal, $._float_literal),
```

Comment:
Can we just keep like how C does it? Just a single block would be preferred to encapsulate both cases

Reply:
Integer and float literals have pretty different rules, which is why I split them up. Trying to do them in one block is probably why they're currently inaccurate...

Reply:
if they're hidden it wouldn't matter though, in the main number block you can just return a choice of either

Reply:
Oh I see what you're suggesting.